added feature: `allowClear` in input field

Suyog241005 opened the pull request:
### Reasons for making this change

This PR adds support for a new ui:allowClear flag.

-Updated all theme packages to support allowClear feature

Fixes [#4671](https://github.com/rjsf-team/react-jsonschema-form/issues/4671)


### Checklist

- [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://rjsf-team.github.io/react-jsonschema-form/docs/contributing) of the Markdown text I've added
- [ ] **I'm adding or updating code**
  - [ ] I've added and/or updated tests. I've run `npx nx run-many --target=build --exclude=@rjsf/docs && npm run test:update` to update snapshots, if needed.
  - [ ] I've updated [docs](https://rjsf-team.github.io/react-jsonschema-form/docs) if needed
  - [ ] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/main/CHANGELOG.md) with a description of the PR
- [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

heath-freenome commented on packages/chakra-ui/src/BaseInputTemplate/BaseInputTemplate.tsx:
```diff
@@ -13,6 +13,7 @@ import {
 
 import { Field } from '../components/ui/field';
 import { getChakra } from '../utils';
+import { X } from 'lucide-react';
```
Can you move this up just below the `@rjsf/utils` import. We like to keep the local imports separate from the library imports.

heath-freenome commented on packages/utils/src/types.ts:
```diff
@@ -1083,6 +1083,8 @@ type UIOptionsBaseType<T = any, S extends StrictRJSFSchema = RJSFSchema, F exten
     /** If submitButtonOptions is provided it should match the `UISchemaSubmitButtonOptions` type */
     submitButtonOptions?: UISchemaSubmitButtonOptions;
+    /** Flag, if set to `true`, will allow the input to be cleared */
+    allowClear?: boolean;
```
```suggestion
    /** Flag, if set to `true`, will allow the text input fields to be cleared */
    allowClearTextInputs?: boolean;
```

Suyog241005 commented:
> @Suyog241005 Sorry for the delay just got back from vacation on Monday and I've been busy. First of all, thank you for being willing to add new features to RJSF, we really appreciate it.
> 
> I've given you some general feedback on the approach you took and a few high-level requests for changes.

@heath-freenome
Thanks for the review and for the detailed feedback.

I’ve gone through your comments and I’m actively reworking the implementation to better align with RJSF’s architecture (templates, localization via translateString, and theme-specific button handling).

Suyog241005 commented:
@heath-freenome I’ve pushed updates addressing the feedback:

- Renamed the UI option to `ui:allowClearTextInputs`

- Added ClearButton via theme-specific templates where applicable

- Updated handlers to use useCallback 

- Integrated localization via translateString

- Fixed import ordering and removed unused types

Please let me know if anything needs further adjustment.

heath-freenome commented on packages/antd/src/templates/BaseInputTemplate/index.tsx:
```diff
@@ -10,6 +10,7 @@ import {
   RJSFSchema,
   StrictRJSFSchema,
 } from '@rjsf/utils';
+import ClearButton from '../IconButton';
```
Thanks for making a separate button... Now you'll need to get it directly from the registry rather than importing it. This needs to be done in all the themes

Something like:
```
  const { ClearButton } = registry.templates.ButtonTemplates;

```

heath-freenome reviewed (commented):
@Suyog241005 Almost there. Thanks for responding to my previous feedback. Sorry for the slow response, between the end-of-year rush at work plus the holidays I haven't had much time to look over these PRs

Suyog241005 commented:
@heath-freenome I’ve updated the implementation based on the feedback:
- Removed the explicit `aria-label` from the clear button
- Switched to using `ClearButton` from `registry.templates.ButtonTemplates`
- Updated `ClearButton` enum value to **"Clear input"** for proper localization usage
- Adjusted the `ClearButton` className in the core theme for consistency

Please let me know if this aligns with the expected direction or if further tweaks are needed.

heath-freenome commented on packages/core/src/components/templates/ButtonTemplates/IconButton.tsx:
```diff
@@ -63,8 +63,8 @@ export function ClearButton<T = any, S extends StrictRJSFSchema = RJSFSchema, F
       icon='remove'
-      className='btn-add col-xs-12'
-      title={translateString(TranslatableString.AddButton)}
+      className='btn-clear'
```
Did you mean to clear the `col-xs-12`?

Suyog241005 commented:
@heath-freenome Please let me know if there’s anything else I should adjust.

heath-freenome commented on packages/antd/src/templates/BaseInputTemplate/index.tsx:
```diff
@@ -58,6 +59,15 @@ export default function BaseInputTemplate<
   const handleFocus = ({ target }: FocusEvent<HTMLInputElement>) => onFocus(id, target && target.value);
 
+  const handleClear = useCallback(
+    (e: React.MouseEvent) => {
```
Dang it, I wanted to merge this when I noticed that you explicitly used `React.MouseEvent` rather than importing it above, in between ChangeEvent and FocusEvent.

heath-freenome commented on packages/chakra-ui/src/BaseInputTemplate/BaseInputTemplate.tsx:
```diff
@@ -39,15 +39,25 @@ export default function BaseInputTemplate<
   const _onBlur = ({ target }: FocusEvent<HTMLInputElement>) => onBlur(id, target && target.value);
   const _onFocus = ({ target }: FocusEvent<HTMLInputElement>) => onFocus(id, target && target.value);
+  const onClear = useCallback(
+    (e: React.MouseEvent) => {
```
Ditto on `React.MouseEvent`

heath-freenome commented on packages/core/src/components/templates/BaseInputTemplate.tsx:
```diff
@@ -73,6 +73,16 @@ export default function BaseInputTemplate<
+    [onChange, options.emptyValue],
+  );
+
+  const { ClearButton } = registry.templates.ButtonTemplates;
```
Consider moving this up below the `} = props;` destructure to keep it closer to the `registry` declaration.

heath-freenome commented on packages/daisyui/src/templates/BaseInputTemplate/BaseInputTemplate.tsx:
```diff
@@ -74,29 +75,45 @@ export default function BaseInputTemplate<
+    [onChange, options.emptyValue],
+  );
+
+  const { ClearButton } = registry.templates.ButtonTemplates;
```
Ditto on moving this up.

heath-freenome commented on packages/mantine/src/templates/BaseInputTemplate.tsx:
```diff
@@ -75,6 +76,17 @@ export default function BaseInputTemplate<
+    [onChange, options.emptyValue],
+  );
+
+  const { ClearButton } = registry.templates.ButtonTemplates;
```
Ditto on moving this up

heath-freenome commented on packages/mui/src/BaseInputTemplate/BaseInputTemplate.tsx:
```diff
@@ -65,6 +67,15 @@ export default function BaseInputTemplate<
       }
     : InputLabelProps;
+  const _onClear = useCallback(
+    (e: React.MouseEvent) => {
```
Ditto on `React.MouseEvent`

heath-freenome commented on packages/react-bootstrap/src/BaseInputTemplate/BaseInputTemplate.tsx:
```diff
@@ -42,6 +43,15 @@ export default function BaseInputTemplate<
   const _onBlur = ({ target }: FocusEvent<HTMLInputElement>) => onBlur(id, target && target.value);
   const _onFocus = ({ target }: FocusEvent<HTMLInputElement>) => onFocus(id, target && target.value);
+  const _onClear = useCallback(
+    (e: React.MouseEvent) => {
```
Ditto on `React.MouseEvent`

heath-freenome commented on packages/react-bootstrap/src/BaseInputTemplate/BaseInputTemplate.tsx:
```diff
@@ -42,6 +43,15 @@ export default function BaseInputTemplate<
+    },
+    [onChange, options.emptyValue],
+  );
+  const { ClearButton } = registry.templates.ButtonTemplates;
```
Ditto on moving this up

heath-freenome commented on packages/semantic-ui/src/BaseInputTemplate/BaseInputTemplate.tsx:
```diff
@@ -55,6 +55,15 @@ export default function BaseInputTemplate<
   const _onBlur = () => onBlur && onBlur(id, value);
   const _onFocus = () => onFocus && onFocus(id, value);
+  const _onClear = useCallback(
+    (e: React.MouseEvent) => {
```
Ditto on `React.MouseEvent`

heath-freenome commented on packages/shadcn/src/BaseInputTemplate/BaseInputTemplate.tsx:
```diff
@@ -51,6 +52,15 @@ export default function BaseInputTemplate<
   const _onBlur = ({ target }: FocusEvent<HTMLInputElement>) => onBlur(id, target && target.value);
   const _onFocus = ({ target }: FocusEvent<HTMLInputElement>) => onFocus(id, target && target.value);
+  const _onClear = useCallback(
+    (e: React.MouseEvent) => {
```
Ditto on `React.MouseEvent`

Suyog241005 commented on packages/primereact/src/BaseInputTemplate/BaseInputTemplate.tsx:
```diff
@@ -37,6 +37,15 @@ export default function BaseInputTemplate<
   } = props;
 
+  const _onClear = useCallback(
+    (e: React.MouseEvent) => {
```
I could also move this block:

```ts
if (Array.isArray(schema.examples)) {
  return <AutoCompleteWidget {...props} />;
}
```
directly above the main `return`, which would allow keeping hooks unconditionally at the top while preserving the current behavior. Let me know if you prefer that approach.
